src/UnitCell.cpp

```diff
@@ -9,261 +9,245 @@
+Vector3D UnitCell::lengths() const {
+    switch (shape_) {
+    case INFINITE:
+        return {0, 0, 0};
+    case ORTHORHOMBIC:
+        return {matrix_[0][0], matrix_[1][1], matrix_[2][2]};
+    case TRICLINIC:
+        break;  // See below
     }
-    a_ = val;
-    update_matrix();
+
+    Vector3D v1 = {matrix_[0][0], matrix_[1][0], matrix_[2][0]};
+    Vector3D v2 = {matrix_[0][1], matrix_[1][1], matrix_[2][1]};
+    Vector3D v3 = {matrix_[0][2], matrix_[1][2], matrix_[2][2]};
+
+    return {v1.norm(), v2.norm(), v3.norm()};
 }
 
-void UnitCell::set_b(double val) {
-    if (shape_ == INFINITE) {
-        throw error("can not set 'b' on infinite cell");
+Vector3D UnitCell::angles() const {
+    switch (shape_) {
+    case INFINITE:
+    case ORTHORHOMBIC:
+        return {90, 90, 90};
+    case TRICLINIC:
+        break;  // See below
     }
-    b_ = val;
-    update_matrix();
+    Vector3D v1 = {matrix_[0][0], matrix_[1][0], matrix_[2][0]};
+    Vector3D v2 = {matrix_[0][1], matrix_[1][1], matrix_[2][1]};
+    Vector3D v3 = {matrix_[0][2], matrix_[1][2], matrix_[2][2]};
+
+    return {
+        rad2deg(acos(dot(v2, v3) / (v2.norm() * v3.norm()))),
+        rad2deg(acos(dot(v1, v3) / (v1.norm() * v3.norm()))),
+        rad2deg(acos(dot(v1, v2) / (v1.norm() * v2.norm())))
+    };
 }
 
-void UnitCell::set_c(double val) {
+void UnitCell::set_lengths(Vector3D lengths) {
     if (shape_ == INFINITE) {
-        throw error("can not set 'c' on infinite cell");
+        throw error("can not set lengths for an infinite cell");
     }
-    c_ = val;
-    update_matrix();
-}
 
-void UnitCell::set_alpha(double val) {
-    if (shape_ != TRICLINIC) {
-        throw error("can not set 'alpha' on non triclinic cell");
+    if (lengths[0] <= 0 || lengths[1] <= 0 || lengths[2] <= 0 ||
+        is_roughly_zero(lengths[0]) || is_roughly_zero(lengths[1]) || is_roughly_zero(lengths[2])) {
+        throw error("can not set unit cell lengths to negative values");
     }
-    alpha_ = val;
-    update_matrix();
-}
 
-void UnitCell::set_beta(double val) {
-    if (shape_ != TRICLINIC) {
-        throw error("can not set 'beta' on non triclinic cell");
+    if (!is_upper_triangular(matrix_)) {
+        warning("UnitCell", "resetting unit cell orientation in set_lengths");
     }
-    beta_ = val;
-    update_matrix();
+
+    // Reset the unit cell, and remove any existing rotation.
+    *this = UnitCell(std::move(lengths), this->angles());
 }
 
-void UnitCell::set_gamma(double val) {
+void UnitCell::set_angles(Vector3D angles) {
     if (shape_ != TRICLINIC) {
-        throw error("can not set 'gamma' on non triclinic cell");
+        throw error("can not set angles for a non-triclinic cell");
+    }
+
+    if (angles[0] <= 0 || angles[1] <= 0 || angles[2] <= 0 ||
+        is_roughly_zero(angles[0]) || is_roughly_zero(angles[1]) || is_roughly_zero(angles[2])) {
+        throw error("can not set unit cell angles to negative values");
```

Comment:
The error message is somewhat inconsistent: [in the constructor](https://github.com/chemfiles/chemfiles/blob/ca56e4d2a922e05558c25132509977c804c4300a/src/UnitCell.cpp#L64-L70) values of zero and negative values are differentiated. Would it be better to define a (inlined) helper function something like `check_angles` that could be used everywhere to throw consistent errors? Same for lengths.

Reply:
Sounds good!